docs/design/vsock-design.md

```diff
@@ -0,0 +1,208 @@
+wouldn't be compliant with the Firecracker security model - a new attack
+surface would be added, providing the guest with a potential path to break
+isolation. 
+
+At the same time, VirtIO vsock is the most prevalent communication mechanism
+employed by today's container orchestrators - an ecosystem with which
+Firecracker aims to integrate, as a secure runtime, backed by hardware
+virtualization.
+
+## Proposed Approach
+
+Our proposed solution is to provide full virtio-vsock support to software
+running inside the guest VM, while bypassing vhost kernel code on the host. To
+that end, Firecracker will implement the virtio-vsock device model, and mediate
+communication between AF_UNIX sockets (on the host end) and AF_VSOCK sockets
+(on the guest end). Both guest and host software will be able to use the
+familiar socket interface.
+
+In order to provide channel multiplexing, AF_VSOCK ports will be translated
+into multiple AF_UNIX sockets (one unix socket per vsock port). Each vsock
+device will have a unix socket path associated. E.g. `/path/to/vsock_{CID}`.
+There are two scenarios to be considered, depending on where the connection is
+initiated.
+
+### Guest-Initiated Connections
+
+When the virtio-vsock device model in Firecracker detects a connection
+request coming from the guest (a VIRTIO_VSOCK_OP_REQUEST packet), it will try
+to forward the connection to an AF_UNIX socket listening on the host, at
+`/path/to/vsock_{CID}:{PORT}`, where `{PORT}` is the destination port, as
+specified in the connection request packet. If no such socket exists, or
+no one is listening on it, a connection cannot be established, and
+a VIRTIO_VSOCK_OP_RST packet will be sent back to the guest.
+
+From the user perspective, these would be the steps taken to establish a
+communication channel:
+
+1. Host: At VM configuration time, add a virtio-vsock device with `CID` and
+   `PATH`;
+2. Host: create and listen on an AF_UNIX socket at `{PATH}_{CID}:{PORT}`;
+3. Guest: create an AF_VSOCK socket and issue a `connect()` call to `HOST_CID`
+   and `PORT`;
+4. Host: `accept()` the new connection.
+
+The channel is established between the sockets obtained at steps 4 (host)
+and 3 (guest).
+
+### Host-Initiated Connections
+
+Firecracker will be listening on an AF_UNIX socket, at `{PATH}_{CID}`. When
+the host needs to initiate a connection, it will connect to that Unix socket,
+then send a special *connect* data packet on it, specifying the destination
+AF_VSOCK port. Following that, the same connection will be forwarded by
+Firecracker to the guest software listening on that port, thus establishing
+the requested channel. If no one is listening, Firecracker will terminate the
+host connection.
+
+1. Host: At VM configuration time, add a virtio-vsock device with `CID` and
+   `PATH`;
+2. Guest: create an AF_VSOCK socket and `listen()` on `PORT`;
+3. Host: `connect()` to AF_UNIX at `{PATH}_{CID}`;
+4. Host: `send()` *connect {PORT}* data packet on that connection;
```

Comment:
That's really putting some burden on the user shoulder here unfortunately. We will expect each host program to send this extra request. This would be worth developing some kind of package in some languages (Rust, Golang) for this.

Reply:
For the similar HyperKit implementation we implemented: https://github.com/linuxkit/virtsock/blob/master/pkg/vsock/hyperkit_darwin.go

This is a generic Go wrapper for virtio vsock and hyper-v sockets. Happy to extend this to support the firecracker interface too.